src/main/scala/lms/core/stub.scala

```diff
@@ -1257,6 +1267,29 @@ abstract class DslDriverCPP[A: Manifest, B: Manifest] extends DslDriverC[A, B] w
   compilerCommand = "g++ -std=c++17 -O3"
 }
 
+abstract class CompilerC[A:Manifest, B:Manifest] extends DslDriverC[A, B] { q =>
+
+  // get original graph
+  val (typeMap, g) = Adapter.genGraph1(manifest[A], manifest[B])(x => Unwrap(wrapper(Wrap[A](x))))
+
+  // FIMXE(feiw) typeMap should be updated during transformation
+  // run some transformation
+  val g1: Graph = (new AdapterTransformer).transform(g)
+  val g2: Graph = (new AdapterTransformer).transform(g1)
+
+  // codegen
+  override lazy val (code, statics) = {
+    val source = new java.io.ByteArrayOutputStream()
+    val statics = time("codegen") {
+      codegen.typeMap = typeMap
+      codegen.stream = new java.io.PrintStream(source)
+      codegen.emitAll(g2, "Snippet")(manifest[A], manifest[B])
+      codegen.extractAllStatics.toList
+    }
+    (source.toString, statics)
+  }
+}
```

Comment:
An example of compiler class that allows users to add several transformation passes before codegen :) 

I am adding 2 same transformations as a show case. Maybe should revise

Importantly, it is still not accurate because we are not updating the typeMap. It is OK for now because the transformer added is `identity transformation`. But this needs to be fixed too

Reply:
AdapterTransformer could be just an object, so here we can just write `AdapterTransformer.transform(g)`.

Ideally, I think the composition of transformers is just function composition. But the typeMap effect breaks this compositionality so that we need to take additional care of typeMap.

Reply:
I don't think AdapterTransformer should be an object because I still want to extend it into other transformers such that different transformation passes can be built.

I do need to think about how typeMap works (not sure it is correct now). But as long as we reuse Adapter object, we will operation on typeMap. Somehow I need to save the old typeMap and reset it to a new one.